### PR TITLE
Add jpipe_link decorator for explicit function binding to pipeline

### PR DESCRIPTION
**Pull request by mosser**
This pull request introduces the new `@jpipe_link` decorator, which allows explicit binding of Python functions to pipeline element IDs, removing the need for fragile label-to-function-name mapping. The changes include the implementation of the decorator, integration into the runtime and pipeline engine, and comprehensive unit tests to ensure correct behavior. Additionally, there are improvements to error messages and display formatting.

### New Feature: Explicit Function-to-Element Binding

* Added the `@jpipe_link` decorator (`link_decorator.py`) that allows developers to explicitly associate a function with a pipeline element by its ID, making the mapping robust and decoupled from function names.
* Updated the pipeline execution logic to use the `function_name` set by `@jpipe_link` (instead of sanitized labels) throughout node processing, error reporting, and skip checks (`engine.py`). [[1]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366L420-R468) [[2]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366L431-R479) [[3]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366L486-R534) [[4]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366L527-R581)

### Runtime and Engine Integration

* Implemented `build_link_registry` in `PythonRuntime` to scan loaded modules for functions decorated with `@jpipe_link`, returning a mapping of element IDs to function names.
* Added `_apply_link_registry`, `_qualified_id`, and `_resolve_node_id` methods to `PipelineEngine` to update graph nodes with the correct function bindings before validation and execution. [[1]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366R361-R362) [[2]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366R373-R418)

### Testing

* Added comprehensive unit tests for the new decorator, registry building, and engine integration, covering various usage scenarios and edge cases (`test_link_decorator.py`).

### Other Improvements

* Improved error and log messages to use qualified node IDs, making debugging clearer. [[1]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366L315-R315) [[2]](diffhunk://#diff-c60f89735d3dd77d74906479905ee14e7c2c21c53443f3fc6ad98881d08e2366L527-R581)
* Enhanced the pretty display output to include the justification name in the variable display and improved line truncation logic (`runner.py`).

### Maintenance

* Bumped the package version to `3.4.0` in `pyproject.toml`.
* Minor cleanup (removed unused imports, updated `__init__.py` for decorator exports). [[1]](diffhunk://#diff-ac0621dc17c3fb4cfd6c13823f9adf1b64d116f381bb865378cd7f365b2de9b9L15) [[2]](diffhunk://#diff-2ae28a8d74266ba9ffdd45eeca7dc5433963df2af6088f6c896770db8155a871R1-R3) [[3]](diffhunk://#diff-5b3efd5683b60d10191b714b2b765adbad28ea1be4cce1d306661b91bdc61f1fR8)

These changes make function-to-element mapping more robust, improve maintainability, and provide thorough test coverage for the new behavior.…s by id

Replaces the fragile label-sanitisation mapping with an explicit binding: @jpipe_link("justification:element_id") on a Python function pins it to the corresponding JSON element regardless of function name or label wording.

- New decorator jpipe_link() sets __jpipe_link_id__ on the function and propagates correctly through @jpipe / @skip / @contribution in any stacking order (functools.wraps copies __dict__)
- PythonRuntime.build_link_registry() scans loaded modules for linked fns
- PipelineEngine._apply_link_registry() / _resolve_node_id() resolve both plain ("e1") and qualified ("performant:e1") id formats before validation
- All user-facing node references now use the fully qualified id via _qualified_id() helper (error messages, skip logs, cycle reports)
- Report display uses justification_name:element_id format and a single truncated line per node (fixes ANSI-inflated status-bar width calculation)
- jpipe_link re-exported from jpipe_decorator for backward compatibility

### CODE REVIEW

File: src/jpipe_runner/runner.py [hunk 1]
```diff
@@ -173,18 +172,15 @@ def pretty_display(diagrams: Iterable[tuple[str, Iterable[dict]]]) -> tuple[int,
...
-            wrapped_label_lines = textwrap.wrap(full_line, width=width - status_bar_len - 1)
-
-            for i, line in enumerate(wrapped_label_lines):
-                if i == 0:
-                    print(line.ljust(width - status_bar_len) + status_bar)
-                else:
-                    print(line.ljust(width))
+            max_content = width - status_bar_visual_len - 1
+            if len(full_line) > max_content:
+                full_line = full_line[: max_content - 3] + "..."
```
[the remark this answers is not in the document]
Comment (mosser) [reply src/jpipe_runner/runner.py]: Fixed. Added guards for `max_content <= 0` (empty string) and `max_content <= 3` (clipped `"..."`) before the standard truncation path. Committed in a84c314.

File: src/jpipe_runner/framework/decorators/link_decorator.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        @jpipe_link("E1")
+        @jpipe(consume=["file_path"], produce=["file_exists"])
+        def check_file(file_path, produce):
+            produce("file_exists", os.path.isfile(file_path))
+            return True
+    """
+
+    def decorator(func: Callable) -> Callable:
+        func.__jpipe_link_id__ = element_id
+        return func
```
[the remark this answers is not in the document]
Comment (mosser) [reply src/jpipe_runner/framework/decorators/link_decorator.py]: Fixed. Introduced a `TypeVar F` bound to `Callable` and changed the signature to `Callable[[F], F]`, so type checkers preserve the wrapped function's type. Committed in a84c314.

File: tests/unit/test_link_decorator.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
+import importlib
+import json
+import os
+import sys
+import tempfile
+import unittest
+
+import networkx as nx
+
```
[the remark this answers is not in the document]
Comment (mosser) [reply tests/unit/test_link_decorator.py]: Fixed. Removed `importlib`, `sys`, and `networkx` from the test file imports. Committed in a84c314.

File: src/jpipe_runner/runtime.py [hunk 4]
```diff
@@ -115,6 +115,22 @@ def call_function(self, name: str, *args, **kwargs) -> Any:
...
+        for every callable decorated with @jpipe_link.
+
+        :return: Dict mapping element_id → attr_name in the loaded modules.
+        :rtype: dict[str, str]
+        """
+        registry = {}
+        for module in self._modules:
+            for attr_name in dir(module):
+                obj = getattr(module, attr_name, None)
+                if callable(obj) and (eid := getattr(obj, "__jpipe_link_id__", None)):
```
[the remark this answers is not in the document]
Comment (mosser) [reply src/jpipe_runner/runtime.py:130]: Fixed. `build_link_registry()` now raises a `RuntimeException` when the same element id is found on two different callables, making duplicate bindings an explicit error rather than a silent overwrite. Committed in a84c314.

File: src/jpipe_runner/framework/engine.py [hunk 5]
```diff
@@ -312,7 +312,7 @@ def get_execution_order(self) -> list[str]:
                 cycle = None
 
             if cycle:
-                cycle_labels = [self.graph.nodes[n].get("label", n) for n in cycle]
+                cycle_labels = [self.graph.nodes[n].get("label", self._qualified_id(n)) for n in cycle]
```
[the remark this answers is not in the document]
Comment (mosser) [reply src/jpipe_runner/framework/engine.py]: Fixed. Each cycle entry is now formatted as `"Label (justification:node_id)"` when a label is present, falling back to just the qualified id when not. This makes cycle diagnostics unambiguous even when labels are non-unique. Committed in a84c314.
